### PR TITLE
Reduce memory footprint by reordering struct elements

### PR DESCRIPTION
**Pull request by florianl**
## What does this PR do?

Reorder struct elements to reduce the memory footprint. Reducing memory footprint also have a direct impact on the GC as less work is needed on GC side.

## Why is it important?

This PR improves the performance of the Elastic Agent without changing its behavior or business logic.

## Checklist

- [ ] ~~My code follows the style guidelines of this project~~
PR does not introduce code changes.
- [ ] ~~I have commented my code, particularly in hard-to-understand areas~~
PR does not introduce code changes.
- [ ] ~~I have made corresponding changes to the documentation~~
PR does not introduce code changes.
- [ ] ~~I have made corresponding change to the default configuration files~~
No configuration changes with this PR.
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
PR does not introduce code changes.
- [ ] ~~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~
PR does not introduce code changes.

### CODE REVIEW

File: internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go [hunk 1]
```diff
@@ -26,11 +26,11 @@ type stateStore interface {
 // Unenroll results in  running agent entering idle state, non managed non standalone.
 // For it to be operational again it needs to be either enrolled or reconfigured.
 type Unenroll struct {
```
Comment (florianl) [inline internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go:28]: struct with 72 pointer bytes could be 56

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 2]
```diff
@@ -44,15 +44,15 @@ var (
 
 // Upgrader performs an upgrade
 type Upgrader struct {
```
Comment (florianl) [inline internal/pkg/agent/application/upgrade/upgrade.go:46]: struct with 120 pointer bytes could be 96

File: internal/pkg/agent/operation/operation_retryable.go [hunk 3]
```diff
@@ -18,8 +18,8 @@ import (
 // if nth operation fails all preceding are retried as well
 type retryableOperations struct {
```
Comment (florianl) [inline internal/pkg/agent/operation/operation_retryable.go:19]: struct with 40 pointer bytes could be 24

File: internal/pkg/agent/stateresolver/resolve.go [hunk 4]
```diff
@@ -60,10 +60,10 @@ func (s *state) String() string {
 }
 
 type active struct {
```
Comment (florianl) [inline internal/pkg/agent/stateresolver/resolve.go:62]: struct with 272 pointer bytes could be 256

File: internal/pkg/composable/providers/kubernetes/pod.go [hunk 5]
```diff
@@ -23,15 +23,15 @@ import (
 )
 
 type pod struct {
```
Comment (florianl) [inline internal/pkg/composable/providers/kubernetes/pod.go:25]: struct with 120 pointer bytes could be 104

File: internal/pkg/composable/providers/kubernetesleaderelection/config.go [hunk 6]
```diff
@@ -8,10 +8,12 @@ import "github.com/elastic/elastic-agent-autodiscover/kubernetes"
 
 // Config for kubernetes_leaderelection provider
 type Config struct {
```
Comment (florianl) [inline internal/pkg/composable/providers/kubernetesleaderelection/config.go:10]: struct with 40 pointer bytes could be 24

File: internal/pkg/core/plugin/process/app.go [hunk 7]
```diff
@@ -35,34 +35,32 @@ var (
 
 // Application encapsulates a concrete application ran by elastic-agent e.g Beat.
 type Application struct {
```
Comment (florianl) [inline internal/pkg/core/plugin/process/app.go:37]: struct with 296 pointer bytes could be 264

File: internal/pkg/core/state/state.go [hunk 8]
```diff
@@ -76,9 +76,9 @@ func FromProto(s proto.StateObserved_Status) Status {
 // State wraps the process state and application status.
 type State struct {
```
Comment (florianl) [inline internal/pkg/core/state/state.go:77]: struct with 40 pointer bytes could be 24

File: internal/pkg/core/status/reporter.go [hunk 9]
```diff
@@ -39,19 +39,19 @@ func (s AgentStatusCode) String() string {
 
 // AgentApplicationStatus returns the status of specific application.
 type AgentApplicationStatus struct {
```
Comment (florianl) [inline internal/pkg/core/status/reporter.go:41]: struct with 64 pointer bytes could be 48

File: internal/pkg/core/status/reporter.go [hunk 10]
```diff
@@ -39,19 +39,19 @@ func (s AgentStatusCode) String() string {
...
 	ID      string
 	Name    string
-	Status  state.Status
 	Message string
-	Payload map[string]interface{}
+	Status  state.Status
 }
 
 // AgentStatus returns the overall status of the Elastic Agent.
 type AgentStatus struct {
```
Comment (florianl) [inline internal/pkg/core/status/reporter.go:50]: struct with 72 pointer bytes could be 48

File: internal/pkg/core/status/reporter.go [hunk 11]
```diff
@@ -68,15 +68,15 @@ type Controller interface {
 }
 
 type controller struct {
```
Comment (florianl) [inline internal/pkg/core/status/reporter.go:70]: struct with 104 pointer bytes could be 88

File: internal/pkg/core/status/reporter.go [hunk 12]
```diff
@@ -272,15 +272,15 @@ type Reporter interface {
 }
 
 type reporter struct {
```
Comment (florianl) [inline internal/pkg/core/status/reporter.go:274]: struct with 80 pointer bytes could be 48

File: internal/pkg/crypto/io.go [hunk 13]
```diff
@@ -21,11 +21,11 @@ import (
 // Option is the default options used to generate the encrypt and decrypt writer.
 // NOTE: the defined options need to be same for both the Reader and the writer.
 type Option struct {
```
Comment (florianl) [inline internal/pkg/crypto/io.go:23]: struct with 40 pointer bytes could be 8

File: internal/pkg/fleetapi/acker/retrier/retrier.go [hunk 14]
```diff
@@ -32,19 +32,19 @@ type Option func(*Retrier)
 
 // Retrier implements retrier for actions acks
 type Retrier struct {
```
Comment (florianl) [inline internal/pkg/fleetapi/acker/retrier/retrier.go:34]: struct with 96 pointer bytes could be 48

File: internal/pkg/reporter/fleet/reporter.go [hunk 15]
```diff
@@ -38,12 +38,12 @@ func (e *event) Message() string {
 
 // Reporter is a reporter without any effects, serves just as a showcase for further implementations.
 type Reporter struct {
```
Comment (florianl) [inline internal/pkg/reporter/fleet/reporter.go:40]: struct with 88 pointer bytes could be 56

File: internal/pkg/reporter/fleet/reporter.go [hunk 16]
```diff
@@ -70,7 +70,7 @@ func (r *Reporter) Report(ctx context.Context, e reporter.Event) error {
 	r.queue = append(r.queue, &event{
 		AgentID:   r.info.AgentID(),
 		EventType: e.Type(),
-		Ts:        fleetapi.Time(e.Time()),
```
Comment (florianl) [inline internal/pkg/reporter/fleet/reporter.go:73]: silent the linter: `ST1003: struct field Ts should be TS (stylecheck)`

File: internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go [hunk 17]
```diff
@@ -75,7 +75,9 @@ func (h *Unenroll) Handle(ctx context.Context, a fleetapi.Action, acker store.Fl
 	} else if h.stateStore != nil {
 		// backup action for future start to avoid starting fleet gateway loop
 		h.stateStore.Add(a)
-		h.stateStore.Save()
```
Comment (florianl) [inline internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go:78]: linter complained: Error return value of `h.stateStore.Save` is not checked (errcheck)

File: internal/pkg/crypto/io.go [hunk 18]
```diff
@@ -189,13 +188,17 @@ func (w *Writer) writeBlock(b []byte) error {
 		return w.err
 	}
 
-	w.writer.Write(iv)
```
Comment (florianl) [inline internal/pkg/crypto/io.go:192]: linter complained: Error return value of `a.monitor.Cleanup` is not checked (errcheck)

File: internal/pkg/crypto/io.go [hunk 19]
```diff
@@ -189,13 +188,17 @@ func (w *Writer) writeBlock(b []byte) error {
...
+	if _, err := w.writer.Write(iv); err != nil {
+		return errors.Wrap(err, "failed to write IV")
+	}
 
 	encodedBytes := w.gcm.Seal(nil, iv, b, nil)
 
 	l := make([]byte, 4)
 	binary.LittleEndian.PutUint32(l, uint32(len(encodedBytes)))
-	w.writer.Write(l)
```
Comment (florianl) [inline internal/pkg/crypto/io.go:198]: linter complained: Error return value of a.monitor.Cleanup is not checked (errcheck)

File: internal/pkg/crypto/io.go [hunk 20]
```diff
@@ -325,7 +328,7 @@ func (r *Reader) consumeBlock() error {
 	}
 
 	encodedBytes := make([]byte, l)
-	_, err = io.ReadAtLeast(r.reader, encodedBytes, int(l))
```
Comment (florianl) [inline internal/pkg/crypto/io.go:328]: linter complained: unnecessary conversion (unconvert)

File: internal/pkg/core/plugin/process/app.go [hunk 21]
```diff
@@ -277,6 +275,7 @@ func (a *Application) setState(s state.Status, msg string, payload map[string]in
 }
 
 func (a *Application) cleanUp() {
+	// nolint:errcheck
 	a.monitor.Cleanup(a.desc.Spec(), a.pipelineID)
```
Comment (florianl) [inline internal/pkg/core/plugin/process/app.go:279]: linter complained: Error return value of `a.monitor.Cleanup` is not checked (errcheck)

File: internal/pkg/core/plugin/process/app.go [hunk 22]
```diff
@@ -231,6 +228,7 @@ func (a *Application) watch(ctx context.Context, p app.Taggable, proc *process.I
 		a.setState(state.Restarting, msg, nil)
 
 		// it was a crash
+		// nolint:errcheck
 		a.start(ctx, p, cfg, true)
```
Comment (florianl) [inline internal/pkg/core/plugin/process/app.go:232]: linter complained: Error return value of `a.start` is not checked (errcheck)

**Review by AndersonQ — CHANGES_REQUESTED** [review]
Whereas I do appreciate adding these errors handling to errors that were ignored before, I understand it might have a bigger impact on the agent's beheviour. @ph do you know this code paths? i'm not familiar with them, I cannot really evaluate the impact.

Another thing, would you mind adding a comment on the structs you aligned the fields to avoid someone reorganizing them for whatever reason?

File: internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go (hunk 17 above)
Comment (florianl) [reply internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go:78]: do you prefer a `// nolint: errcheck` instruction?

Comment (ph) [reply internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go:78]: Lets do the nolint here, I am not sure of the impact of this changes here, wdyt @michel-laterman ?

**Review by ph — CHANGES_REQUESTED** [review]
@florianl Lets do the changes for the struct memory alignment but lets not changes the error handling or linter errors.

File: internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go (hunk 17 above)
Comment (michel-laterman) [reply internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go:78]: I think that currently `state.Action` can be either unenroll or policy-change. (this is different from the action queue part I added).

I think if the action fails to persist, then the agent will try to start with the persisted policy if it's restarted. But at that point it's api key has likely been revoked.

Comment (florianl) [reply internal/pkg/agent/application/pipeline/actions/handlers/handler_action_unenroll.go:78]: updated the PR to use the `nolint` instruction.

**Review by ph — APPROVED** [review]

**Review by AndersonQ — APPROVED** [review]
I'm approving to unblock the PR, but without a comment on the structs to let others know that the fields are aligned, I believe there is a good chance someone will mess up the aliment

**Comment by ph** [comment]
@AndersonQ Good point, it think it more practice.

@florianl did you an extension or you just manually reviewed all of them? I wonder if this could be added on the ci?

**Comment by ph** [comment]
Considering the memory, it's great but I don't think must of the change in this PR really reduce the overall memory footprint, most of the struct are long lived.

**Comment by florianl** [comment]
@ph I used [`fieldalignment`](https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/fieldalignment) which is part of the Go tools collection.

**Review by ph — APPROVED** [review]

**Pull request merged**